experiment/selinux/src/selinux.rs

```diff
@@ -44,268 +95,301 @@ impl SELinux {
-    pub fn ro_file_label() {
-        unimplemented!("not implemented yet")
-    }
+        // check if selinuxfs is available before going the slow path
+        let fs = fs::read_to_string("/proc/filesystems").unwrap_or_default();
+        if !fs.contains("\tselinuxfs\n") {
+            return None;
+        }
 
-    // function similar with kvmContainerLabels in go-selinux repo.
-    // kvm_container_labels returns the default processLabel and mountLabel to be used
-    // for kvm containers by the calling process.
-    pub fn kvm_container_labels() {
-        unimplemented!("not implemented yet")
+        // continue reading until finding mount point
+        loop {
+            // slow path: try to find among the mounts
+            match File::open("/proc/self/mountinfo") {
+                Ok(file) => {
+                    let reader = BufReader::new(file);
+                    for line in reader.lines().map_while(Result::ok) {
+                        if let Some(mnt) = Self::find_selinux_fs_mount(&line) {
+                            if Self::verify_selinux_fs_mount(mnt.clone()) {
+                                return Some(mnt);
+                            }
+                        }
+                    }
+                }
+                Err(_) => return None,
+            }
+        }
     }
 
-    // function similar with initContainerLabels in go-selinux repo.
-    // init_container_labels returns the default processLabel and file labels to be
-    // used for containers running an init system like systemd by the calling process.
-    pub fn init_container_labels() {
-        unimplemented!("not implemented yet")
+    // This function returns the path to the mountpoint of an selinuxfs
+    // filesystem or an empty string if no mountpoint is found. Selinuxfs is
+    // a proc-like pseudo-filesystem that exposes the SELinux policy API to
+    // processes. The existence of an seliuxfs mount is used to determine
+    // whether SELinux is currently enabled or not.
+    pub fn get_selinux_mountpoint(&mut self) -> Option<&PathBuf> {
+        // Avoiding code conflicts and ensuring thread-safe execution once only.
+        if !self.selinuxfs_init_done.load(Ordering::SeqCst) {
+            self.selinuxfs = Self::find_selinux_fs();
+            self.selinuxfs_init_done.store(true, Ordering::SeqCst);
+        }
+        self.selinuxfs.as_ref()
     }
 
-    // function similar with containerLabels in go-selinux repo.
-    // container_labels returns an allocated processLabel and fileLabel to be used for
-    // container labeling by the calling process.
-    pub fn container_labels() {
-        unimplemented!("not implemented yet")
+    // classIndex returns the int index for an object class in the loaded policy,
+    // or -1 and an error.
+    // If class is "file" or "dir", then return the corresponding index for selinux.
+    pub fn class_index(&mut self, class: &str) -> Result<i64, SELinuxError> {
```

Comment:
Can `class` be represented by enums or something?

Reply:
If the number of class is not so big, I think this is a good idea.

However, seems that there are more than 100 classes.
So I don't think it is a good idea implementing class as enums or something.

I confirmed the number of classes on amazon linux environment using the command `seinfo -c`.
There are 133 classes.